### PR TITLE
Create RTCDataChannel to send microphone activities message

### PR DESCRIPTION
**Pull request by niuyunsi**
TODO(yunsi): Add UI to show microphone activity for remote stream

### CODE REVIEW

File: src/components/Conference.tsx [hunk 1]
```diff
@@ -303,18 +336,66 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+            const receiveChannel = event.channel;
+            receiveChannel.onmessage = (messageEvent) => { this.handleReceiveChannelMessage(messageEvent, id) };
+            this.receiveChannels[id] = receiveChannel;
+        }
+    }
+
+    private handleReceiveChannelMessage(event: MessageEvent, id: string) {
+        if (event.data) {
+            switch (event.data) {
+                case 'speaking':
```
Comment (asinner) [inline src/components/Conference.tsx]: Define interfaces for our data channel protocol and use a discriminated union type here

File: src/components/Conference.tsx [hunk 2]
```diff
@@ -191,9 +199,29 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+        })
+
         this.connection.subscribe(this.handleIncomingMessage);
     }
 
+    // NOTE(yunsi): Send the speaking message to all clients through data channels
+    private broadcastDataChannelMessage(message: string) {
+        Object.keys(this.sendChannels).forEach((id: string) => {
+            const sendChannel = this.getSendChannelById(id);
+            if (sendChannel.readyState === 'open') {
```
Comment (asinner) [inline src/components/Conference.tsx]: Create a function `sendToDataChannel(id: string, message: DataChannelMessage)` that encapsulates business logic for readyState

File: src/components/Conference.tsx [hunk 3]
```diff
@@ -191,9 +199,29 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
         this.setState({
             localStream: stream,
         })
+
+        // NOTE(yunsi): Add an audio monitor to listen to the speaking change of local stream.
+        let audioMonitor = Hark(this.state.localStream);
```
Comment (asinner) [inline src/components/Conference.tsx]: state is not set yet use `Hark(stream)`

File: src/components/Conference.tsx [hunk 4]
```diff
@@ -191,9 +199,29 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+        audioMonitor.on('speaking', () => {
+            this.broadcastDataChannelMessage('speaking')
```
Comment (asinner) [inline src/components/Conference.tsx]: use a function to produce a message for the protocol

File: src/components/Conference.tsx [hunk 5]
```diff
@@ -191,9 +199,29 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+        })
+        audioMonitor.on('stopped_speaking', () => {
+            this.broadcastDataChannelMessage('stopped_speaking')
```
Comment (asinner) [inline src/components/Conference.tsx]: use a function to produce a message for the protocol

File: src/components/Conference.tsx [hunk 6]
```diff
@@ -248,18 +276,23 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
     }
 
     private createPeerConnectionById(id: string) {
-        // TODO(yunsi): Add RTCPeerConnection config.
         const peerConnection = new RTCPeerConnection(this.props.peerConnectionConfig);
+        // TODO(yunsi): Add data channel config
+        const sendChannel = peerConnection.createDataChannel('dataChannel');
```
Comment (asinner) [inline src/components/Conference.tsx]: Talked about send/receive channels in person

File: src/components/Conference.tsx [hunk 7]
```diff
@@ -303,18 +336,66 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+        const receiveChannel = this.getReceiveChannelById(id);
 
         if (peerConnection) {
             peerConnection.close();
         } else {
             console.warn('handleRemovePeerMessage(): Missing connection Id: %s', id);
         }
 
+        if (sendChannel) {
+            sendChannel.close();
```
Comment (asinner) [inline src/components/Conference.tsx]: consolidate if possible

File: src/components/Conference.tsx [hunk 8]
```diff
@@ -303,18 +336,66 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+        } else {
+            console.warn('handleRemovePeerMessage(): Missing sendChannel Id: %s', id);
+        }
+
+        if (receiveChannel) {
+            receiveChannel.close();
```
Comment (asinner) [inline src/components/Conference.tsx]: consolidate if possible

File: src/components/Conference.tsx [hunk 9]
```diff
@@ -46,17 +47,23 @@ const userMediaConfig = {
...
 }
 
+// TODO(yunsi): Add data channel config
+// var dataChannelConfig: RTCDataChannelInit = {}
+
 export interface IConferenceState {
     localId: ConfUserID | undefined;
     localStream: MediaStream | undefined;
     remoteStreams: { [id: string]: MediaStream };
+    remoteIsSpeaking: { [id: string]: boolean };
```
Comment (asinner) [inline src/components/Conference.tsx]: Is this being used anywhere in the render?

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
+interface RTCDataChannelInit {
```
Comment (asinner) [inline index.d.ts]: Move all this to custom.d.ts

File: src/components/Conference.tsx [hunk 11]
```diff
@@ -191,9 +199,29 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+        })
+        audioMonitor.on('stopped_speaking', () => {
+            this.broadcastDataChannelMessage('stopped_speaking')
+        })
+
         this.connection.subscribe(this.handleIncomingMessage);
     }
 
+    // NOTE(yunsi): Send the speaking message to all clients through data channels
+    private broadcastDataChannelMessage(message: string) {
```
Comment (asinner) [inline src/components/Conference.tsx]: use a protocol for the message

File: src/components/Conference.tsx [hunk 12]
```diff
@@ -303,18 +336,66 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+        }
+
+        if (receiveChannel) {
+            receiveChannel.close();
+        } else {
+            console.warn('handleRemovePeerMessage(): Missing receiveChannel Id: %s', id);
+        }
+
         delete this.peerConnections[id];
+        delete this.sendChannels[id];
```
Comment (asinner) [inline src/components/Conference.tsx]: consolidate if possible

File: src/components/Conference.tsx [hunk 13]
```diff
@@ -414,4 +495,12 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
     private getPeerConnectionById(id: string): RTCPeerConnection {
         return this.peerConnections[id]
     }
+
+    private getSendChannelById(id: string): RTCDataChannel {
```
Comment (asinner) [inline src/components/Conference.tsx]: consolidate if possible

File: src/components/Conference.tsx (hunk 9 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Currently we just store this information, but we need to add UI to show microphone activity for remote stream based on this.state.remoteIsSpeaking and also support custom rendering for microphone activity.

File: src/components/Conference.tsx (hunk 3 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 4 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 5 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 11 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 2 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 6 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: src/components/Conference.tsx (hunk 1 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

File: index.d.ts (hunk 10 above)
Comment (niuyunsi) [reply index.d.ts]: Done

File: src/components/Conference.tsx [hunk 14]
```diff
@@ -188,12 +201,40 @@ export class Conference extends React.Component<IConferenceProps, IConferenceSta
...
+    // NOTE(yunsi): Send the speaking message to all clients through data channels
+    private broadcastDataChannelMessage(message: IDataChannelMessage) {
+        Object.keys(this.dataChannels).forEach((id: string) => {
+            this.sendMessageToDataChannel(message, id)
+        });
+    }
+
+    private sendMessageToDataChannel(message: IDataChannelMessage, id: string) {
+        const dataChannel = this.getDataChannelById(id);
+        if (dataChannel && dataChannel.readyState === DataChannelReadyState.OPEN) {
```
Comment (asinner) [inline src/components/Conference.tsx]:
```
if (!dataChannel) {
  console.log(warning);
  return
}

if (dataChannel.readyState !== DataChannelReadyState.OPEN) {
  console.log(warning);
  return
}
```

**Review by asinner — APPROVED** [review]
Approved - test out the data channel JSON messages to be sure

File: src/components/Conference.tsx (hunk 14 above)
Comment (niuyunsi) [reply src/components/Conference.tsx]: Done

**Review by asinner — APPROVED** [review]

**Pull request merged**